src/G2/Equiv/Verifier.hs

```diff
@@ -270,9 +372,48 @@ verifyLoop' solver ns_pair sh1 sh2 prev =
         S.UNSAT () -> return $ Just (not_ready_h)
         _ -> return Nothing
 
-getObligations :: State t -> State t -> Maybe [Obligation]
-getObligations s1 s2 =
-  case proofObligations s1 s2 (exprExtract s1) (exprExtract s2) of
+{-
+Algorithm:
+Perform the sub-expression extraction in here
+The prev list is fully expanded already
+Compare the sub-expressions to the prev state pairs
+If any match occurs, try extrapolating it
+If extrapolation works, we can flag the real state pair as a repeat
+-}
+induction :: (HS.HashSet Name, HS.HashSet Name) ->
+             [(StateET, StateET)] ->
+             (StateET, StateET) ->
+             Bool
+induction ns_pair prev (s1, s2) =
+  let s1' = inductionState s1
+      s2' = inductionState s2
+      prev' = filter statePairInduction prev
+      prev'' = map (\(p1, p2) -> (inductionState p1, inductionState p2)) prev'
+      hm_maybe_list = map (mrHelper' ns_pair (Just HM.empty) (s1', s2')) prev''
+      -- try matching the prev state pair with other prev state pairs
+      hm_maybe_zipped = zip hm_maybe_list prev'
+      -- ignore the combinations that didn't work
+      hm_maybe_zipped' = [(hm, p) | (Just hm, p) <- hm_maybe_zipped]
+      csp = concretizeStatePair (expr_env s1, expr_env s2)
+      concretized = map (uncurry csp) hm_maybe_zipped'
+      -- TODO optimization:  just take the first one
+      concretized' = if null concretized then [] else [head concretized]
```

Comment:
Maybe change to a direct pattern match (using a `case` statement) on `concretized`.